Plugin did not respond

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

<!--- Please keep this note for the community --->

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

<!--- Thank you for keeping this note for the community --->


### Terraform Version

v1.7.2

### AzureRM Provider Version

v3.113.0

### Affected Resource(s)/Data Source(s)

azurerm_network_security_rule

### Terraform Configuration Files

```hcl
#############################################################################
Main.tf
#############################################################################
data "azurerm_client_config" "current" {}
data "azurerm_subscription" "current" {}

locals {
  common_tags = {
    Owner       = var.tag_owner
    function    = var.tag_function
    CreatedDate = timestamp()
    CreatedBy   = var.tag_CreatedBy
  }
}

# Generate random number value
resource "random_string" "number" {
  length  = 4
  special = false
  lower   = false
  numeric = true
  upper   = false
}

resource "azurerm_resource_group" "dc_rg" {
  name     = var.dc_rg_name
  location = var.location
  tags     = merge(local.common_tags)
  lifecycle {
    ignore_changes = [
      tags["CreatedDate"]
    ]
  }
}

resource "azurerm_resource_group" "vnet_rg" {
  name     = var.vnet_rg_name
  location = var.location
  tags     = merge(local.common_tags)
  lifecycle {
    ignore_changes = [
      tags["CreatedDate"]
    ]
  }
}

#region Create VNET and multiple Subnets and NSGs + NSGs_Rules ##########

resource "azurerm_virtual_network" "vnet" {
  name                = var.vnet
  location            = azurerm_resource_group.vnet_rg.location
  resource_group_name = azurerm_resource_group.vnet_rg.name
  address_space       = var.address_space
  dns_servers         = var.dns_servers
}

resource "azurerm_subnet" "subs" {
  for_each             = var.subnets
  name                 = each.key
  resource_group_name  = azurerm_resource_group.vnet_rg.name
  virtual_network_name = azurerm_virtual_network.vnet.name
  address_prefixes     = each.value["address_prefixes"]
}

resource "azurerm_network_security_group" "nsg" {
  name                = var.nsg
  location            = var.location
  resource_group_name = azurerm_resource_group.vnet_rg.name
}

# Associate the NSG with the Subnet
resource "azurerm_subnet_network_security_group_association" "sub_nsg" {
  for_each                  = azurerm_subnet.subs
  depends_on                = [azurerm_network_security_group.nsg]
  subnet_id                 = azurerm_subnet.subs[each.key].id
  network_security_group_id = azurerm_network_security_group.nsg.id
}

# resource "azurerm_network_security_rule" "nsg_rules" {
#   for_each               = var.nsg_rules
#   name                   = each.key
#   priority               = each.value["priority"]
#   direction              = each.value["direction"]
#   access                 = each.value["access"]
#   protocol               = each.value["protocol"]
#   source_port_range      = each.value["source_port_range"]
#   destination_port_range = each.value["destination_port_range"]
#   #source_address_prefixes     = each.value["source_address_prefixes"]
#   destination_address_prefix  = each.value["destination_address_prefix"]
#   resource_group_name         = azurerm_resource_group.vnet_rg.name
#   network_security_group_name = azurerm_network_security_group.nsg.name # azurerm_network_security_group.nsgs[each.value["network_security_group_name"]].name
# }

resource "azurerm_network_security_rule" "rule" {
  count                       = length(var.security_rules)
  name                        = var.security_rules[count.index].name
  priority                    = var.security_rules[count.index].priority
  direction                   = var.security_rules[count.index].direction
  access                      = var.security_rules[count.index].access
  protocol                    = var.security_rules[count.index].protocol
  source_address_prefix       = var.security_rules[count.index].source_address_prefix
  source_port_ranges          = var.security_rules[count.index].source_port_ranges
  destination_address_prefix  = var.security_rules[count.index].destination_address_prefix
  destination_port_range      = var.security_rules[count.index].destination_port_range # Accessing the first element
  resource_group_name         = azurerm_resource_group.vnet_rg.name
  network_security_group_name = azurerm_network_security_group.nsg.name
}

#endregion #########################################

#create a Storage Account
resource "azurerm_storage_account" "sa" {
  for_each                 = var.storage_accounts
  name                     = "stavmdiag${each.key}"
  location                 = azurerm_resource_group.dc_rg.location
  resource_group_name      = azurerm_resource_group.dc_rg.name
  account_tier             = each.value["account_tier"]
  account_replication_type = each.value["account_replication_type"]
}

resource "azurerm_key_vault" "kvvaultpass" {
  name                       = "kvpass${random_string.number.result}"
  location                   = azurerm_resource_group.dc_rg.location
  resource_group_name        = azurerm_resource_group.dc_rg.name
  tenant_id                  = data.azurerm_client_config.current.tenant_id
  soft_delete_retention_days = 7
  purge_protection_enabled   = false
  sku_name                   = "standard"
  depends_on = [
    azurerm_resource_group.dc_rg
  ]
}

resource "azurerm_key_vault_access_policy" "client" { // This is for AD Users Logged into Azure to give them the right access when creating resources. 
  key_vault_id        = azurerm_key_vault.kvvaultpass.id
  tenant_id           = data.azurerm_client_config.current.tenant_id
  object_id           = data.azurerm_client_config.current.object_id
  secret_permissions  = ["Backup", "Delete", "Get", "List", "Purge", "Recover", "Restore", "Set", ]
  key_permissions     = ["Backup", "Create", "Decrypt", "Delete", "Encrypt", "Get", "Import", "List", "Purge", "Recover", "Restore", "Sign", "UnwrapKey", "Update", "Verify", "WrapKey", ]
  storage_permissions = ["Backup", "Delete", "DeleteSAS", "Get", "GetSAS", "List", "ListSAS", "Purge", "Recover", "RegenerateKey", "Restore", "Set", "SetSAS", "Update", ]
  depends_on = [
    azurerm_key_vault.kvvaultpass
  ]
}

resource "random_password" "vm_admin_password" {
  length  = 12
  special = true
}

resource "azurerm_key_vault_secret" "vm_admin_password_secret" {
  name         = "vm-admin-password"
  value        = random_password.vm_admin_password.result
  key_vault_id = azurerm_key_vault.kvvaultpass.id
}


#region separate dc-vm module

resource "azurerm_network_interface" "nic" {
  for_each            = var.servers
  name                = "${each.key}-nic"
  location            = var.location
  resource_group_name = azurerm_resource_group.dc_rg.name

  ip_configuration {
    name                          = "internal"
    subnet_id                     = azurerm_subnet.subs[each.value["subnet"]].id
    private_ip_address_allocation = "Dynamic"
    #public_ip_address_id          = (each.value["public_ip"] != false ? azurerm_public_ip.pips[each.key].id : null) ## if public ip var parameters are true, catch the value 
  }
}

resource "azurerm_managed_disk" "datadisks" {
  for_each             = var.datadisks
  name                 = each.key
  location             = var.location
  resource_group_name  = azurerm_resource_group.dc_rg.name
  storage_account_type = each.value["storage_account_type"]
  create_option        = each.value["create_option"]
  disk_size_gb         = each.value["disk_size_gb"]
}

resource "azurerm_windows_virtual_machine" "Servers" {
  for_each            = var.servers
  name                = each.key
  resource_group_name = azurerm_resource_group.dc_rg.name
  location            = var.location
  computer_name       = each.value["computer_name"]
  size                = each.value["size"]
  admin_username      = each.value["admin_username"]
  admin_password      = azurerm_key_vault_secret.vm_admin_password_secret.value #data.azurerm_key_vault_secret.admin_pass.value ## Thats the value from the secret in the azure portal
  network_interface_ids = [
    azurerm_network_interface.nic[each.key].id
  ]

  #delete_os_disk_on_termination = true
  #delete_data_disks_on_termination = true

  os_disk {
    caching              = each.value["caching"]
    storage_account_type = each.value["storage_account_type"]
  }

  source_image_reference {
    publisher = each.value["publisher"]
    offer     = each.value["offer"]
    sku       = each.value["sku"]
    version   = each.value["version"]
  }

  # resources are generated here
  boot_diagnostics {
    storage_account_uri = azurerm_storage_account.sa["diag"].primary_blob_endpoint
  }

}

resource "azurerm_virtual_machine_data_disk_attachment" "diskattach" {
  for_each = var.datadisks
  managed_disk_id    = azurerm_managed_disk.datadisks[each.key].id
  virtual_machine_id = azurerm_windows_virtual_machine.servers[each.key].id
  lun                = each.value["lun"]
  caching            = each.value["caching"]
}

#endregion main.tf ##########################################################


#############################################################################
variables.tf
#############################################################################

variable "location" {
  type    = string
  default = "westus"
}

variable "tenant_id" {
  type    = string
  default = "00000000-0000-0000-0000-000000000000"
}

variable "dc_rg_name" {
  type        = string
  description = "RG name"
}

# variable "vnet_identity_name" {
#   type        = string
#   description = "VNET Identity name"
# }

variable "tag_function" {
  type        = string
  description = "Please provide details on the function that the resource is assigned to."
}

variable "tag_owner" {
  type        = string
  description = "Tag Owner"
}

variable "tag_CreatedBy" {
  type        = string
  description = "Tag Created By"
}

variable "vnet_rg_name" {
  type        = string
  description = "VNET RG Name"
}

#Variable for VNET Address
variable "address_space" {
  type    = list(string)
  default = ["10.0.0.0/27"]
}

variable "vnet" {
  type        = string
  description = "VNET Name"
}

variable "dns_servers" {
  description = "The DNS servers to be used with vNet"
  type        = list(string)
  default     = []
}

variable "subnets" {
  type = map(any)
  default = {
    snet-iu-it-dc-we1-prd-1 = {
      address_prefixes = ["10.0.0.0/28"]
    }
    snet-iu-it-dc-we1-prd-2 = {
      address_prefixes = ["10.0.0.0/28"]
    }
  }
}
variable "nsg" {
  type        = string
  description = "Network Security Group Name"
}

variable "security_rules" {
  description = "List of security rules"
  type = list(object({
    name                       = string
    priority                   = number
    direction                  = string
    access                     = string
    protocol                   = string
    source_address_prefix      = string
    source_port_ranges         = list(string)
    destination_address_prefix = string
    destination_port_range     = list(string) # This should be a list to handle multiple ports
  }))
  default = []
}

# The parameters Servers are pair with the id subnet created before
variable "servers" {
  type = map(any)
  default = {
    "testvm1" = {
      subnet               = "subnet1"
      computer_name        = "testvm1"
      size                 = "Standard_D4as_v5"
      admin_username       = "adminuser"
      caching              = "ReadWrite"
      storage_account_type = "Standard_LRS"
      publisher            = "MicrosoftWindowsServer"
      offer                = "WindowsServer"
      sku                  = "2022-Datacenter"
      version              = "latest"
      public_ip            = false
    }
    testvm2 = {
      subnet               = "subnet2"
      computer_name        = "testvm2"
      size                 = "Standard_D4as_v5"
      admin_username       = "adminuser"
      caching              = "ReadWrite"
      storage_account_type = "Standard_LRS"
      publisher            = "MicrosoftWindowsServer"
      offer                = "WindowsServer"
      sku                  = "2022-Datacenter"
      version              = "latest"
      public_ip            = false
    }
  }
}

# variable "datadisks" {
#   type = map(any)
#   default = {
#     Disk1 = {
#       storage_account_type = "Standard_LRS"
#       create_option        = "Empty"
#       disk_size_gb         = "30"
#       lun                  = "10"
#       caching              = "ReadWrite"
#       attach_to            = "testvm1"
#     }
#     Disk2 = {
#       storage_account_type = "Standard_LRS"
#       create_option        = "Empty"
#       disk_size_gb         = "30"
#       lun                  = "11"
#       caching              = "ReadWrite"
#       attach_to            = "testvm2"
#     }
#   }
# }

# variable "delete_os_disk_on_termination" {
#   type = bool
# }
# variable "delete_data_disks_on_termination" {
#   type = bool
# }

variable "storage_accounts" {
  type = map(any)
  default = {
    diag = {
      account_tier             = "Standard"
      account_replication_type = "LRS"
    }
    # Scripts = {
    #   account_tier             = "Standard"
    #   account_replication_type = "ZRS"
    # }
  }
}

variable "user_principal_name" {
  type    = string
  default = "contoso@contoso.com"
}


#endregion variables.tf ##########################################################

#############################################################################
config.auto.tfvars.tf
#############################################################################

tag_owner     = ""
tag_CreatedBy = ""
tag_function  = ""
nsg           = ""
dc_rg_name    = ""
vnet          = ""
vnet_rg_name  = ""

security_rules = [
  {
    name                       = "allow-SQL"
    priority                   = 100
    direction                  = "Inbound"
    access                     = "Allow"
    protocol                   = "Udp"
    source_address_prefix      = "*"
    source_port_ranges         = ["*"]
    destination_address_prefix = "*"
    destination_port_range     = "1433"
  },
  {
    name                       = "allow-http"
    priority                   = 101
    direction                  = "Inbound"
    access                     = "Allow"
    protocol                   = "Udp"
    source_port_ranges         = ["*"]
    source_address_prefix      = "*"
    destination_address_prefix = "*"
    destination_port_range     = "1444"
  }
]
```


### Debug Output/Panic Output

```shell
Error: Plugin did not respond
│
│   with azurerm_network_security_rule.rule[1],
│   on main.tf line 92, in resource "azurerm_network_security_rule" "rule":
│   92: resource "azurerm_network_security_rule" "rule" {
│
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may contain more details.
╵

Stack trace from the terraform-provider-azurerm_v3.113.0_x5.exe plugin:

panic: interface conversion: interface {} is string, not *schema.Set

goroutine 61 [running]:
github.com/hashicorp/terraform-provider-azurerm/internal/services/network.resourceNetworkSecurityRuleUpdate(0xc005184e80, {0x72b1800?, 0xc001b3e900?})
        github.com/hashicorp/terraform-provider-azurerm/internal/services/network/network_security_rule_resource.go:323 +0x1b56
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).update(0x8ac81e0?, {0x8ac81e0?, 0xc0051b3dd0?}, 0xd?, {0x72b1800?, 0xc001b3e900?})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.29.0/helper/schema/resource.go:800 +0x163
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc0018ff500, {0x8ac81e0, 0xc0051b3dd0}, 0xc0009c9860, 0xc005184d00, {0x72b1800, 0xc001b3e900})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.29.0/helper/schema/resource.go:919 +0x83a
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc000534648, {0x8ac81e0?, 0xc0051b3ce0?}, 0xc000367900)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.29.0/helper/schema/grpc_provider.go:1060 +0xdbc
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0xc000a21ae0, {0x8ac81e0?, 0xc0051b34d0?}, 0xc000d75490)
        github.com/hashicorp/terraform-plugin-go@v0.19.0/tfprotov5/tf5server/server.go:859 +0x56b
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0x7f46fa0?, 0xc000a21ae0}, {0x8ac81e0, 0xc0051b34d0}, 0xc000d75420, 0x0)
        github.com/hashicorp/terraform-plugin-go@v0.19.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:467 +0x169
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000026000, {0x8af2ee0, 0xc0011c4820}, 0xc0051bab40, 0xc001787710, 0xe10a818, 0x0)
        google.golang.org/grpc@v1.58.3/server.go:1374 +0xde7
google.golang.org/grpc.(*Server).handleStream(0xc000026000, {0x8af2ee0, 0xc0011c4820}, 0xc0051bab40, 0x0)
        google.golang.org/grpc@v1.58.3/server.go:1751 +0x9e7
google.golang.org/grpc.(*Server).serveStreams.func1.1()
        google.golang.org/grpc@v1.58.3/server.go:986 +0xbb
created by google.golang.org/grpc.(*Server).serveStreams.func1 in goroutine 54
        google.golang.org/grpc@v1.58.3/server.go:997 +0x145

Error: The terraform-provider-azurerm_v3.113.0_x5.exe plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```


### Expected Behaviour

The provider applies the configuration changes, creating or updating resources in Azure without errors.

### Actual Behaviour

The Terraform AzureRM provider plugin encountered a critical error and crashed while attempting to apply the network security rule changes.

### Steps to Reproduce

Running terraform apply resulted in the plugin crashing with a type conversion error.

### Important Factoids

Germany

### References

_No response_

--- a/internal/services/network/network_security_rule_resource.go
+++ b/internal/services/network/network_security_rule_resource.go
@@ -320,12 +320,12 @@ func resourceNetworkSecurityRuleUpdate(d *pluginsdk.ResourceData, meta interface
 
 	if d.HasChange("source_port_ranges") {
 		var sourcePortRanges []string
-		r := d.Get("destination_port_range").(*pluginsdk.Set).List()
+		r := d.Get("source_port_ranges").(*pluginsdk.Set).List()
 		for _, v := range r {
 			s := v.(string)
 			sourcePortRanges = append(sourcePortRanges, s)
 		}
-		payload.Properties.DestinationPortRanges = pointer.To(sourcePortRanges)
+		payload.Properties.SourcePortRanges = pointer.To(sourcePortRanges)
 	}
 
 	if d.HasChange("destination_port_range") {

